### PR TITLE
[JENKINS-34244] Some UX improvements in the New item form.

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-34244](https://issues.jenkins-ci.org/browse/JENKINS-34244)
- [x] Form button control stick to the bottom of the page (_sticky button control_).
- [x] Submit button needs to be three-state: hover, active and inactive.
- [x] Submit button is disabled until the client-side validation is valid.
- [x] Fix tests

@reviewbybees especially @gusreiber since he asked these modifications.
@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by daniel-beck** [comment]
`Initial source code modifications` is neither a useful PR title, nor a useful commit message.

**Comment by recena** [comment]
@daniel-beck You have more information in the Jira ticket 😄 I'll squash the PR you prefer.

**Comment by batmat** [comment]
+1 with @daniel-beck the history should preferably be informative already using `git log`, without having to look into JIRA, though indeed more details there is good. (your second commit is more meaningful seems like e.g.)

File: core/src/main/resources/hudson/model/View/newJob.properties [hunk 1]
```diff
@@ -2,7 +2,7 @@ NewJob=New {0}
 JobName={0} name
 ItemName.help=Required field
 ItemName.label=Enter an item name
-ItemName.validation.required=This field cannot be empty. Please enter a valid name and press OK button.
+ItemName.validation.required=This field cannot be empty. Please enter a valid name, select a item type and and press OK button.
```
Comment (batmat) [inline core/src/main/resources/hudson/model/View/newJob.properties]: duplicated `and` if this isn't a GH glitch?

File: war/src/main/js/add-item.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-var getItems = function(){
+var getItems = function() {
   var d = $.Deferred();
   $.get('itemCategories?depth=3').done(
-      function(data){
-        d.resolve(data);
-      }
+    function(data){
+      d.resolve(data);
+    }
```
Comment (batmat) [inline war/src/main/js/add-item.js]: All whitespace modifications or do I miss something?

**Comment by recena** [comment]
@batmat I agree, it was only a commit bad documented. I'll squash the PR if it is approved.

File: war/src/main/js/add-item.js [hunk 3]
```diff
@@ -82,6 +82,27 @@ $.when(getItems()).done(function(data) {
       $('.input-help', context).removeClass('input-message-disabled');
     }
 
+    var doSticky = function() {
+      var s = $('form .footer .btn-decorator');
```
Comment (batmat) [inline war/src/main/js/add-item.js]: Maybe it's a common pattern in JS, but shouldn't we use a longer/more explicit variable name than `s`?

File: war/src/main/js/add-item.js (hunk 2 above)
Comment (recena) [inline war/src/main/js/add-item.js]: @batmat No, I took advantage this revision of _New Item_ form in order to have the same coding style convention in this file.

File: core/src/main/resources/hudson/model/View/newJob.properties (hunk 1 above)
Comment (recena) [inline core/src/main/resources/hudson/model/View/newJob.properties]: Good catch 😄 I'll be addressed in my next commit.

Comment (daniel-beck) [inline core/src/main/resources/hudson/model/View/newJob.properties]: Grammar: *an* item name

**Comment by recena** [comment]
@batmat, @daniel-beck Some of your comments were addressed here https://github.com/jenkinsci/jenkins/pull/2324/commits/bfce23d019a604d180f26d32c87a8ce91d8d3f56

**Comment by recena** [comment]
@jenkinsci/code-reviewers We need testers 😄

**Comment by recena** [comment]
I'm working on this failed [test](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5051/testReport/hudson.model/MyViewTest/testDoCreateItem/).

**Comment by amuniz** [comment]
Locally tested.

The last point (_Submit button is disabled until the client-side validation is valid_) is not true in the current state. If you set a name the submit button is activated `onblur`, but the client side validation is still invalid as you have to select an item type.

Beyond that, I think a permanent solution for the UX of this form would be to add the action button (_OK_ in this case) on each row right side, using some kind of righ-pointing arrow image (meaning _Continue_ for the user). And of course, remove the current OK button. But this is just my opinion.

**Comment by recena** [comment]
> The last point (Submit button is disabled until the client-side validation is valid) is not true in the current state. If you set a name the submit button is activated onblur, but the client side validation is still invalid as you have to select an item type.

Let me try it. This is not the expected behavior.

> Beyond that, I think a permanent solution for the UX of this form would be to add the action button (OK in this case) on each row right side, using some kind of righ-pointing arrow image (meaning Continue for the user). And of course, remove the current OK button. But this is just my opinion.

Indeed, I can work on it and prepare a proposal.

**Comment by gusreiber** [comment]
@recena So it looks like the OK button isn't pegged to the screen as it is in the following config form page [ https://drive.google.com/file/d/0B6_0AE0-eDhad2Y4OVFGYkdkYVE/view ] As a result, it is easy to get the tab focus off the screen. If the OK button were pegged on the screen, I don't think you would be able to get into the state of an off window focus nearly as easily as I did in the above video.

...the above suggestion of removing the OK button altogether and making it clear that on selection of the item type you will immediately advance to the configuration would be another way to go. The items in the list would likely need some sort of border or frame clarifying that they are buttons. Maybe a right pointing arrow as suggested above would work.

I think restoring the sticky OK button is still the way to go, though, as it most matches both the v1 behavior and matches the subsequent config page, and addresses all the other issues surrounding an off-window focus.

My concern about fully 'buttonizing' the items in the list is that especially as we move more and more to a flat aesthetic, it gets harder and hard to have a clear indication that an element is an action button. To overcome that I worry we will need to add some sort of extra design goop to the page to fully clarify that each one of these links is itself an advancing button. That goops adds clutter and redundancy.

My $.02.

In the meantime, I guess I am 🐛 this because the OK button still falls off the screen, which doesn't match the following config page that has an identical looking Save button. I can be talked out of it, though, as we have killed this horse.

**Comment by recena** [comment]
@gusreiber I watched the screencast (thanks so much) but it seems the screencast does not use this PR.

**Comment by recena** [comment]
@gusreiber you can check it easily because as part of this PR [I've reviewed](https://github.com/jenkinsci/jenkins/pull/2324/files#diff-d3b7548c8366c6d10c3aeb795a6dd1beL5) the validation message for the name of the Item.

**Comment by gusreiber** [comment]
@recena K. yeah, I must have had the wrong branch.
https://drive.google.com/file/d/0B6_0AE0-eDhaVUJjRngyOEUzRkE/view

There is still some issues with the focus after selecting an item. If the page is going to scroll down after selecting, the focus should also go to the OK button. Not sure the page should scroll down after selection, given the OK button is already there, but the focus should go to it.

If you want to avoid the pegging all together and just advance immediately after selection the type, that might make sense, but I would still have the concern I mention above.

**Comment by recena** [comment]
@gusreiber 

> but I would still have the concern I mention above.

What do you mean?

**Comment by recena** [comment]
@gusreiber ping?

**Comment by recena** [comment]
@gusreiber, @amuniz Your comments were addressed [here](https://github.com/jenkinsci/jenkins/pull/2324/commits/e05a32887d3ebff3f9102e913a39a527f140da82). I recommend you to build the WAR locally and check again.

Thanks for your time.

**Comment by recena** [comment]
@jenkinsci/code-reviewers you are always to invite to send your feedback 😄

**Comment by gusreiber** [comment]
@recena minor issue with how the name validation works. It looks like the name field is validated when an item is selected rather than when the input changes or is blurred. As a result, you can remove the name later: https://drive.google.com/file/d/0B6_0AE0-eDhaNDdKRWdXVmRQREE/view 

....pretty minor stuff, but prolly an easy fix, too. sorry to be a pain.  🐝

**Comment by recena** [comment]
@gusreiber It would have been interesting if you Jira issue [JENKINS-34244](https://issues.jenkins-ci.org/browse/JENKINS-34244) describes all these details.

I'm going to modify the current behavior.

**Comment by recena** [comment]
@gusreiber Thanks so much for your feedback, it is great! Your comment was addressed [here](https://github.com/jenkinsci/jenkins/pull/2324/commits/a78637722aea5012ada84ebcb44cc1e14aed3fb0). Could you take a look? I hope the current status is ok for you. I'd like to finish this in the next hours.

**Comment by amuniz** [comment]
:bee: as far as it accomplish the three points in the PR description.

**Comment by recena** [comment]
@amuniz This PR tries to address the @gusreiber's recommendations. I think the PR description summarizes the main goals. Thanks for your feedback.

File: war/src/main/js/add-item.js [hunk 4]
```diff
@@ -82,6 +78,65 @@ $.when(getItems()).done(function(data) {
       $('.input-help', context).removeClass('input-message-disabled');
     }
 
+    // About Scroll-linked effect: https://developer.mozilla.org/en-US/docs/Mozilla/Performance/Scroll-linked_effects
+    function doSticky() {
+      var decorator = $('form .footer .btn-decorator');
```
Comment (kzantow) [inline war/src/main/js/add-item.js]: :ant: It's a preference (and a commonly used convention) to use $ to start names for any variables containing the result of a jQuery selector

File: war/src/main/js/widgets/add/addform.less [hunk 5]
```diff
@@ -23,14 +23,14 @@
 
     .btn-decorator {
       position: absolute;
-      bottom: -1px;
-      left: -1px;
+      bottom: 0px;
       padding: 20px 30px;
       width: auto;
       border-top-right-radius: 10px;
       border: 1px solid #aaccbb;
```
Comment (kzantow) [inline war/src/main/js/widgets/add/addform.less]: NIT: *I* would prefer to add `border-width: 1px 1px 0 0;`

File: war/src/main/js/add-item.js [hunk 6]
```diff
@@ -82,6 +78,65 @@ $.when(getItems()).done(function(data) {
...
+        }
+      });
+    }
+
+    function enableSubmit(status) {
+      var btn = $('form .footer .btn-decorator button[type=submit]');
+      if (status === true) {
+        if (btn.hasClass('disabled')) {
+          btn.removeClass('disabled');
+          btn.attr('disabled', false);
```
Comment (kzantow) [inline war/src/main/js/add-item.js]: :ant: this should be `.prop('disabled'`

File: war/src/main/js/add-item.js [hunk 7]
```diff
@@ -82,6 +78,65 @@ $.when(getItems()).done(function(data) {
...
+        }
+        btn.focus();
+      } else {
+        if (!btn.hasClass('disabled')) {
+          btn.addClass('disabled');
+          btn.attr('disabled', true);
```
Comment (kzantow) [inline war/src/main/js/add-item.js]: ditto

**Comment by kzantow** [comment]
I think the focus behavior is still very bad and is being used to work around a better flow of just moving the name below the item type (and probably hidden until item type is clicked). Anyway, seems to work correctly from my testing so :bee:

File: war/src/main/js/widgets/add/addform.less (hunk 5 above)
Comment (recena) [inline war/src/main/js/widgets/add/addform.less]: Me too but we are using the this design in job configuration. I can push another PR to change both.

**Comment by kzantow** [comment]
Still a :bee: from me :)

**Comment by recena** [comment]
@kzantow Could you file a Jira issue or send me an email with your proposal about the focus behavior?

**Comment by gusreiber** [comment]
🐝

**Comment by recena** [comment]
@reviewbybees done

**Comment by recena** [comment]
@daniel-beck @oleg-nenashev IMO this PR is ready to merge for next weekly release.

**Comment by recena** [comment]
It would be great if we get this PR merged before to start [JENKINS-34952](https://issues.jenkins-ci.org/browse/JENKINS-34952) in order avoid a hell merge.

**Pull request merged**